### PR TITLE
Use host's rng when communication is enabled

### PR DESCRIPTION
**Pull request by christianpoveda**
This uses the host's randomness when the communication enabled flag is used. I am not sure about the error handling. I was thinking about fallbacking to `rand` if `getrandom` fails and also print something so the user knows miri is not using the host's rng because it failed. Let me know what you think. 

Related issue: https://github.com/rust-lang/miri/issues/800.

r? @RalfJung @oli-obk

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -35,6 +35,7 @@ byteorder = { version = "1.1", features = ["i128"]}
 cargo_metadata = { version = "0.8", optional = true }
 directories = { version = "2.0", optional = true }
 rustc_version = { version = "0.2.3", optional = true }
+getrandom = "0.1.10"
```
Comment (oli-obk) [inline Cargo.toml]: getrandom docs say

> This is a low-level API. Most users should prefer using high-level random-number library like rand.

And we already have `rand` in our dependencies. There's probably an appropriate function in the `rand` crate for filling a buffer

Comment (RalfJung) [reply Cargo.toml]: TBH I feel like `getrandom` is the better match here. We are not "most users", we want to model system-level randomness APIs. That is *exactly* what the getrandom crate does.

File: src/helpers.rs [hunk 2]
```diff
@@ -97,9 +97,18 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
         )?.expect("we already checked for size 0");
 
-        let rng = this.memory_mut().extra.rng.get_mut();
         let mut data = vec![0; len];
-        rng.fill_bytes(&mut data);
+
+        if this.machine.communicate {
+            // Fill the buffer using the host's rng.
+            getrandom::getrandom(&mut data).map_err(|err| {
+                InterpError::Unsupported(UnsupportedOpInfo::Unsupported(err.to_string()))
```
Comment (RalfJung) [inline src/helpers.rs]: "Unsupported" does not seem right, does it?

Also we have `err_unsup!` so that you don't need to give these long paths.

Comment (christianpoveda) [reply src/helpers.rs]: What error do you suggest?

**Comment by RalfJung** [comment]
Also could you expand the `-Zmiri-enable-communication` docs to mention this? (Can be in a separate PR.)

**Comment by christianpoveda** [comment]
> Also could you expand the `-Zmiri-enable-communication` docs to mention this? (Can be in a separate PR.)
Sure

File: src/helpers.rs (hunk 2 above)
Comment (RalfJung) [reply src/helpers.rs]: Well I guess "unsupported" is fine, but please don't just forward the error -- prefix something like "getrandom failed: " or so.

Comment (RalfJung) [reply src/helpers.rs]: Actually `err_unsup_format!` seems like the best fit here.

Comment (christianpoveda) [reply src/helpers.rs]: I will fix this and expand the docs for the flag

File: README.md [hunk 3]
```diff
@@ -158,8 +158,8 @@ Several `-Z` flags are relevant for Miri:
   will miss bugs in your program.  However, this can also help to make Miri run
   faster.
 * `-Zmiri-enable-communication` enables communication between the host
-  environment and Miri, i.e., all the host environment variables are available
-  during Miri runtime.
+  environment and Miri, i.e.,Miri uses the host's random number generator and
```
Comment (RalfJung) [inline README.md]:
```suggestion
  environment and Miri, i.e., Miri uses the host's random number generator and 
```

File: src/helpers.rs [hunk 4]
```diff
@@ -2,7 +2,7 @@ use std::mem;
 
 use rustc::ty::{self, layout::{self, Size, Align}};
 use rustc::hir::def_id::{DefId, CRATE_DEF_INDEX};
-use rustc::mir;
+use rustc::{mir, err_unsup_format};
```
Comment (RalfJung) [inline src/helpers.rs]: Is this needed? The `use crate::*` should take care of that...

Comment (christianpoveda) [reply src/helpers.rs]: It does, I just corrected it

**Comment by RalfJung** [comment]
Thanks!

@bors r+

**Pull request merged**